token/program-2022/src/onchain.rs

```diff
@@ -3,19 +3,91 @@
 
 use {
     crate::{
+        error::TokenError,
         extension::{transfer_hook, StateWithExtensions},
         instruction,
         state::Mint,
     },
     solana_program::{
-        account_info::AccountInfo, entrypoint::ProgramResult, instruction::AccountMeta,
-        program::invoke_signed, pubkey::Pubkey,
+        account_info::AccountInfo,
+        entrypoint::ProgramResult,
+        instruction::{AccountMeta, Instruction},
+        msg,
+        program::invoke_signed,
+        program_error::ProgramError,
+        pubkey::Pubkey,
+    },
+    spl_transfer_hook_interface::{
+        error::TransferHookError, get_extra_account_metas_address,
+        onchain::add_cpi_accounts_for_execute,
     },
-    spl_transfer_hook_interface::onchain::add_cpi_accounts_for_execute,
 };
 
+/// Onchain helper to get all additional required account metas for a checked
+/// transfer
+///
+/// Note that this onchain helper will build a new `Execute` instruction,
+/// resolve the extra account metas, and then add them to the transfer
+/// instruction. This is because the extra account metas are configured
+/// specifically for the `Execute` instruction, which requires five accounts
+/// (source, mint, destination, authority, and validation state), wheras the
+/// transfer instruction only requires four (source, mint, destination, and
+/// authority) in addition to `n` number of multisig authorities.
+pub fn resolve_extra_transfer_account_metas_for_cpi<'a>(
+    cpi_instruction: &mut Instruction,
+    cpi_account_infos: &mut Vec<AccountInfo<'a>>,
+    mint_info: &AccountInfo<'a>,
+    additional_accounts: &[AccountInfo<'a>],
+    amount: u64,
+) -> Result<(), ProgramError> {
```

Comment:
Same here, let's pass the transfer parameters by name, rather than in the instruction